File: course_3/module_1/examples/prim_mst.py
import heapq

class Edge:
    """
    Represents an edge in a graph with a source, target, and weight.
    """
    def __init__(self, source, target, weight):
        self.source = source
        self.target = target
        self.weight = weight

    def __repr__(self):
        return f"Edge({self.source} -- {self.weight} --> {self.target})"

    # Comparison methods for heapq to compare edges by weight.
    def __lt__(self, other):
        return self.weight < other.weight

class Graph:
    """
    Represents an undirected weighted graph.
    """
    def __init__(self, num_vertices):
        """
        Initializes the graph with a given number of vertices.
        :param num_vertices: Total number of vertices in the graph.
        """
        self.num_vertices = num_vertices
        self.adjacency_list = {i: [] for i in range(num_vertices)}

    def add_edge(self, source, target, weight):
        """
        Adds an undirected edge to the graph.
        :param source: Starting vertex of the edge.
        :param target: Ending vertex of the edge.
        :param weight: Weight of the edge.
        """
        if source < 0 or source >= self.num_vertices or target < 0 or target >= self.num_vertices:
            raise ValueError("Source or target vertex out of range.")
        edge = Edge(source, target, weight)
        self.adjacency_list[source].append(edge)
        self.adjacency_list[target].append(Edge(target, source, weight))  # Add reverse edge.

    def get_edges(self, vertex):
        """
        Returns all edges connected to a specific vertex.
        :param vertex: The vertex whose edges need to be retrieved.
        :return: A list of edges connected to the vertex.
        """
        return self.adjacency_list[vertex]

class PrimMST:
    """
    Implements Prim's algorithm to find the Minimum Spanning Tree (MST).
    """
    def __init__(self, graph):
        """
        Initializes the PrimMST object with a graph.
        :param graph: An instance of the Graph class.
        """
        self.graph = graph
        self.mst_edges = []  # List to store edges included in the MST.
        self.total_cost = 0  # Total weight of the MST.
        self.visited = []  # Tracks whether a vertex is in the MST.
        self.min_heap = []  # Min-heap to store edges based on their weights.

    def push_edges(self, vertex):
        """
        Pushes all edges of a vertex into the min-heap if they lead to unvisited vertices.
        :param vertex: The vertex whose edges are pushed into the heap.
        """
        self.visited[vertex] = True
        for edge in self.graph.get_edges(vertex):
            if not self.visited[edge.target]:
                heapq.heappush(self.min_heap, edge)

    def find_mst(self, start_vertex=0):
        """
        Finds the Minimum Spanning Tree (MST) starting from a specified vertex.
        :param start_vertex: The vertex to start building the MST from (default is 0).
        :return: A tuple (mst_edges, total_cost), where mst_edges is a list of edges in the MST
                 and total_cost is the sum of their weights.
        """
        self.visited = [False] * self.graph.num_vertices  # Tracks whether a vertex is in the MST.
        self.min_heap = []  # Min-heap to store edges based on their weights.
        self.mst_edges = []
        self.total_cost = 0

        # Push all edges from the start vertex.
        self.push_edges(start_vertex)

        # Main loop: Process edges from the heap.
        while self.min_heap:
            edge = heapq.heappop(self.min_heap)
            if self.visited[edge.target]:
                continue  # Skip edges that lead to already visited vertices.

            # Add the edge to the MST.
            self.mst_edges.append(edge)
            self.total_cost += edge.weight

            # Push edges from the newly added vertex.
            self.push_edges(edge.target)

        return self.mst_edges, self.total_cost

# Example usage:
if __name__ == "__main__":
    # Create a graph with 6 vertices.
    graph = Graph(6)
    graph.add_edge(0, 1, 4)
    graph.add_edge(0, 2, 4)
    graph.add_edge(1, 2, 2)
    graph.add_edge(1, 3, 5)
    graph.add_edge(2, 3, 8)
    graph.add_edge(2, 4, 10)
    graph.add_edge(3, 4, 2)
    graph.add_edge(3, 5, 6)
    graph.add_edge(4, 5, 3)

    # Run Prim's algorithm.
    prim_mst = PrimMST(graph)
    mst_edges, total_cost = prim_mst.find_mst(start_vertex=0)

    # Print the result.
    print("Edges in the MST:")
    for edge in mst_edges:
        print(edge)

    print(f"Total cost of the MST: {total_cost}")

File: course_3/module_1/examples/prim_mst.go
package main

import (
	"container/heap"
	"fmt"
)

// Edge represents an edge in a graph with a source, target, and weight.
type Edge struct {
	Source int
	Target int
	Weight int
}

// Graph represents an undirected weighted graph.
type Graph struct {
	NumVertices int
	AdjList     map[int][]Edge
}

// NewGraph creates a new graph with a specified number of vertices.
func NewGraph(numVertices int) *Graph {
	return &Graph{
		NumVertices: numVertices,
		AdjList:     make(map[int][]Edge),
	}
}

// AddEdge adds an undirected edge to the graph.
func (g *Graph) AddEdge(source, target, weight int) {
	g.AdjList[source] = append(g.AdjList[source], Edge{Source: source, Target: target, Weight: weight})
	g.AdjList[target] = append(g.AdjList[target], Edge{Source: target, Target: source, Weight: weight})
}

// PriorityQueue implements a priority queue for edges.
type PriorityQueue []Edge

func (pq PriorityQueue) Len() int           { return len(pq) }
func (pq PriorityQueue) Less(i, j int) bool { return pq[i].Weight < pq[j].Weight }
func (pq PriorityQueue) Swap(i, j int)      { pq[i], pq[j] = pq[j], pq[i] }

// Push adds an element to the queue.
func (pq *PriorityQueue) Push(x interface{}) {
	*pq = append(*pq, x.(Edge))
}

// Pop removes and returns the smallest element from the queue.
func (pq *PriorityQueue) Pop() interface{} {
	old := *pq
	n := len(old)
	item := old[n-1]
	*pq = old[:n-1]
	return item
}

// PrimMST represents the Prim's algorithm for finding MST.
type PrimMST struct {
	Graph     *Graph
	TotalCost int
	MSTEdges  []Edge
}

// NewPrimMST initializes a new instance of PrimMST.
func NewPrimMST(graph *Graph) *PrimMST {
	return &PrimMST{
		Graph:     graph,
		TotalCost: 0,
		MSTEdges:  []Edge{},
	}
}

// FindMST computes the minimum spanning tree using Prim's algorithm.
func (pmst *PrimMST) FindMST(startVertex int) {
	visited := make([]bool, pmst.Graph.NumVertices)
	pq := &PriorityQueue{}
	heap.Init(pq)

	// Add initial edges from the start vertex.
	pmst.addEdges(startVertex, visited, pq)

	for pq.Len() > 0 {
		// Get the edge with the smallest weight.
		edge := heap.Pop(pq).(Edge)

		// Skip if the target vertex is already visited.
		if visited[edge.Target] {
			continue
		}

		// Include the edge in the MST.
		pmst.MSTEdges = append(pmst.MSTEdges, edge)
		pmst.TotalCost += edge.Weight

		// Add edges from the newly added vertex.
		pmst.addEdges(edge.Target, visited, pq)
	}
}

// addEdges pushes all valid edges from a vertex into the priority queue.
func (pmst *PrimMST) addEdges(vertex int, visited []bool, pq *PriorityQueue) {
	visited[vertex] = true
	for _, edge := range pmst.Graph.AdjList[vertex] {
		if !visited[edge.Target] {
			heap.Push(pq, edge)
		}
	}
}

func main() {
	// Create a graph with 6 vertices.
	graph := NewGraph(6)
	graph.AddEdge(0, 1, 4)
	graph.AddEdge(0, 2, 4)
	graph.AddEdge(1, 2, 2)
	graph.AddEdge(1, 3, 5)
	graph.AddEdge(2, 3, 8)
	graph.AddEdge(2, 4, 10)
	graph.AddEdge(3, 4, 2)
	graph.AddEdge(3, 5, 6)
	graph.AddEdge(4, 5, 3)

	// Run Prim's algorithm.
	primMST := NewPrimMST(graph)
	primMST.FindMST(0)

	// Print the result.
	fmt.Println("Edges in the MST:")
	for _, edge := range primMST.MSTEdges {
		fmt.Printf("%d -- %d (%d)\n", edge.Source, edge.Target, edge.Weight)
	}
	fmt.Printf("Total cost of the MST: %d\n", primMST.TotalCost)
}
